Added automatic naming of "year" & "semester" for data.json

athityakumar opened the pull request.

athityakumar committed:
Added automatic naming of "semester" for data.json

athityakumar committed:
Fixed over-write bug of gsub

athityakumar committed:
Fixed looping error

icyflame commented:
@athityakumar, thanks a lot for this patch.

Correct me if I am wrong. In your changes, essentially, you are seeing the last object in the present `data.json` file, and then, you are changing `mid` and `end`, etc.

The basic **problem** with this approach is that the last paper that was added could very well be `end-autumn-2014`. And since `data.json` is not a sorted list, this assumption will fail.

A better approach to solve this problem is to get the present data and infer the semester, exam and year from it. Most of the cases, we will be right, in the cases that we aren't right, we can simply type the right one in, of course. So, can you re-write this patch to do something of that sort? (Considering that KGP schedules remain almost constant, every year 😄 )

We could also ensure that this list was always sorted, but the issue with that is a lot of contributors do not use this script to add a node, and there is always a possibility that something like that will slip in.

athityakumar commented:
Do you mean that the mid-end spring-autumn conversions should be decided by the Date.today parameter rather than the last object of the data.json? If so, what should be the dates interval for each of the categories - mid spring , end spring, mid autumn , end-autumn?

Or would including a sorting function to sort the data.json inside add-node.rb help the cause?

I'll also try to include batch-insert-semester & batch-insert-year options...

icyflame commented:
Well, these can be the date ranges for now (Obviously, use these as global constants, which are easily editable later):

```
mid-spring - mid February to end of April
end-spring - end April to end of September
mid-autumn - September end to November end
end-autumn - November end to mid-February
```

Batch insert could be a possibility. Let's fix this script first, then, maybe we could do batch insert with a CSV or TXT file of links and subject codes. That could be a possibility too. Great work!

athityakumar commented:
Ok , I'll include the date intervals... I'll send a PR later today regarding this...

Meanwhile, by batch-insert , I meant having AN OPTION of taking batch-year & batch-semester as inputs rather semester & year so that we won't have to input year & semester for each paper on the terminal, during a "Session".

athityakumar committed:
Added batch-insert feature. Updated auto-year & auto-semester feature.

athityakumar commented:
@icyflame : Hey! I added another commit... Can you review the latest version?

icyflame commented on add_node.rb:
```diff
@@ -10,8 +11,37 @@
+MID_SPRING_PERIOD = (this_year+"-02-15 to "+this_year+"-04-30").split(" to ")
+END_SPRING_PERIOD = (this_year+"-05-01 to "+this_year+"-09-30").split(" to ")
+MID_AUTUMN_PERIOD = (this_year+"-10-01 to "+this_year+"-11-30").split(" to ")
+END_AUTUMN_PERIOD = (this_year+"-12-01 to "+(this_year.to_i+1).to_s+"-02-14").split(" to ")
```
So, in these 4 lines, the `split` is unnecessary. Just change this to a list and also, please use the `#{this_year}-MM-dd` notation as that is more intuitive to read than string addition.

icyflame commented on add_node.rb:
```diff
@@ -10,8 +11,37 @@
+	batch = gets.chomp
+	if batch.downcase == "y"
+		puts "Enter batch semester : "
+		batch_semester = gets.chomp
```
Here, `batch_semester` must match the regex for `mid-spring-2016`. This regex is probably there in one of the scripts inside mfqp-source, I will try to find it, you can run `grep -i mid` in the mfqp-source scripts folder and look through the results to find it. I would rather you reuse that regex because that has been checked by myself, and @xtinct as well, in an earlier PR.

icyflame commented on add_node.rb:
```diff
@@ -25,24 +55,43 @@
 
-	puts "Enter semester (mid spring 2016): "
-	semester = gets.chomp
+	if batch_semester.empty?
```
Here, we **must** show the default to the user. So, move the default calculation to the top of the script and show the default here. (Think Linux command defaults.)

icyflame replied:
Hey, this variable will never be empty as it was initialised to a non-empty value. Check the conditional flow up-top and change the initialisation.

icyflame commented:
Hey, thanks for spending the time.  I have added review on individual lines.
1. The year parameter is redundant throughout the script. The year can and should be inferred from the semester value. I made this mistake when I first wrote this script. So, once you are done with the set of changes in the line notes, please do this as well. It will be pretty easy, as far as I can tell. 

Thanks, and good work! :+1:

athityakumar committed:
Updated global constants into arrays

athityakumar committed:
Updated default values. Modified code for redundant variable "year".

athityakumar committed:
Corrected "batch_semester" can't be empty bug

icyflame commented on add_node.rb:
```diff
@@ -10,8 +11,53 @@
+		puts "Enter batch semester (#{default_semester}) : "
+		batch_semester = gets.chomp
+		if batch_semester.length == 0 
+			batch_semester = default_semester
```
Before doing this, make sure that the semester entered by the user matches [this](https://github.com/metakgp/mfqp-source/blob/master/scripts/upload/upload.rb#L18) regex. If not, keep them in the loop, because they probably want batch entry but just entered it wrong. (Remove the ending `pdf` part from the regex on that script)

athityakumar replied:
Ok, so is it like this?

batch_semester : empty => take default_semester
batch_semester : mid/end spring/autumn 2016 => accept given input as valid input
batch_semester : something else => ask for batch_semester again

Should the same thing be done with semester too in case of no batch_semester?

athityakumar committed:
Added looping logic for semester & batch_semester.

athityakumar commented:
@icyflame : Included regex matching for semester & batch_semester... Also converted some part of the code into functions... 😃 

Any other improvement / bug remaining in this code?

raivivek commented on add_node.rb:
```diff
@@ -1,17 +1,69 @@
+			return false
+		end
+	end
+	return true
```
You can replace the whole block with something like:

```
possible_semesters.include? semester
```

athityakumar committed:
Optimized "semester_exists" block

icyflame commented:
@athityakumar Script looks much cleaner now. Good work on the methods, I was going to say it anyway. But I can't see regex matching in the script now. Instead of checking only the first part of the semester, I would rather that you simply copy the whole regex and do a simple `re.match` call, to make sure that we don't let anything not matching the regex for `mid spring 2016` enter `data.json`.

Right now, only the first 10 characters of the semester entered by the user are tested.

raivivek commented on add_node.rb:
```diff
@@ -1,17 +1,67 @@
+
+def semester_exists semester
+	possible_semesters = ["mid spring","end spring","mid autumn","end autumn"]
+	if possible_semesters.include? semester
```
If `semester` exists, why would you want to return `false`? I'm not sure if I understand. May be you wanted something like this:

```
def semester_exists? semester
  possible_semesters = [ .. ]
  possible_semesters.include? semester
end
```

icyflame replied:
@athityakumar This was a quirk I was going to ask you to fix.

@vivekiitkgp He is returning false here, so that the `while` loop condition becomes false.

But I think we all would prefer a much more intuitive method like this:

``` rb
while !semester_exists(semester)
...
```

and inside semester exists, actually return `true` if `semester` is okay, otherwise return `false`.

raivivek commented:
Just wanted to say that you are doing great @athityakumar! Thanks for willing to work on this and promptly addressing comments.

athityakumar committed:
Added regex match & made "semester_exists" function intuitive.

athityakumar commented:
@icyflame : Added the actual regex & made the semester loop logic intuitive. The counter-intuitive one occupied lesser number of characters though... :wink: 

@vivekiitkgp : You're welcome! :smile: 

Any more bug correction / improvement required?

Commit:
Re-indent the file

Commit:
Add editorconfig

Commit:
Add a testing feature

- this will help when testing out whether the script is working as
  intended. Doesn't make changes to anything.

icyflame commented:
@athityakumar 

![screen shot 2016-06-14 at 10 20 23 pm](https://cloud.githubusercontent.com/assets/3668034/16051655/440e26c6-327e-11e6-9c72-0350786b6180.png)

Not sure why you didn't face this problem. I am using ruby 2.0.0.

Please pull from [my new branch](https://github.com/icyflame/mfqp/tree/athityakumar-gh-pages) before fixing that issue. I was under the impression that you had run this script. Please always run scripts before submitting a patch, even if it is a very minor change. Things tend to go astray.

icyflame commented:
Also, @athityakumar I have added `editorconfig` to the repository in those commits on my branch. So, please install the appropriate plugin for your editor, or set your editor to 2-space indents.

athityakumar committed:
Rectified dynamic constant variable allocation bug

athityakumar committed:
Fixed regex syntax error

athityakumar commented:
@icyflame - Pulled your branch & fixed more bugs... Check the latest version , please?

icyflame commented:
@athityakumar Nice work on this patch. Merged it in. 👍 

Just a couple pointers. It seems like the commits you made on `GitHub` were tagged under the username `athithyakumar` rather than your normal git config name of `Athithya Kumar`. I would recommend that you either change your name in the GitHub web interface, or better still, avoid committing on github. It seems convenient, but committing like that without testing isn't a good idea.
